README.md
# rpc/xml

[![build status](https://img.shields.io/travis/kofrasa/rpc.svg)](http://travis-ci.org/kofrasa/rpc) [![GoDoc](https://godoc.org/github.com/kofrasa/rpc?status.svg)](https://godoc.org/github.com/kofrasa/rpc)

XML-RPC codec for Go compatible with [gorilla/rpc](https://github.com/gorilla/rpc)

## install

latest

```sh
go get github.com/kofrasa/rpc
```

versions

```sh
go get gopkg.in/kofrasa/rpc.v1
```

## docs

[godoc](https://godoc.org/github.com/kofrasa/rpc)

## usage

### server

```go

import (
  "net/http"
  "github.com/gorilla/rpc/v2"
  "github.com/kofrasa/rpc/xml"
)

type Args struct {
  A, B int
}

type Reply struct {
  C int
}

type PositionalArgs []int

type Arith int

func (t *Arith) Add(r *http.Request, args *Args, reply *Reply) error {
  reply.C = args.A + args.B
  return nil
}

func (t *Arith) Max(r *http.Request, args *PositionalArgs, reply *Reply) error {
  nums := *args
  reply.C = num[0]
  for i := 1; i < len(num); i++ {
    if reply.C < num[i] {
      reply.C = num[i]
    }
  }
  return nil
}

func main() {
  s := rpc.NewServer()
  s.RegisterCodec(xml.NewServerCodec(), "text/xml")
  s.RegisterService(new(Arith), "Arith")
  http.ListenAndServe("http://localhost:5000", s)
}
```

### client

```go
import (
  "fmt"
  "net/http"
  "time"
  "github.com/kofrasa/rpc/xml"
)

type Args struct {
  A, B int
}

type Reply struct {
  C int
}

const addr = "http://localhost:5000"

func main() {
  client := xml.NewClient(addr)

  args := Args{A: 3, B: 3}
  var reply Reply

  // encodes args as a single struct parameter
  client.Call("Arith.Add", &reply, args)
  fmt.Printf("%d + %d = %d\n", args.A, args.B, reply.C) // expect 6

  // encodes variable arguments as multiple parameters
  client.Call("Arith.Max", &reply, 2, 5, 3)
  fmt.Printf("Max(%d,%d,%d) = %d\n", 2, 5, 3, reply.C) // expect 5

  // you can also create a client with your own custom httpclient
  customHTTPClient := &http.Client{Timeout: time.Second * 5}
  customRPCClient := xml.NewClient(addr, xml.WithHTTPClient(customHTTPClient))
}

```

## features

* Extended [iso8601](https://en.wikipedia.org/wiki/ISO_8601) formats.
  * `20060102T15:04:05`
  * `2006-01-02T15:04:05`
  * `2006-01-02T15:04:05-07:00`
  * `2006-01-02T15:04:05Z07:00`
* Decodes boolean `true` and `false`
* Server method aliases
* Server accept encoding for `gzip` and `deflate`
* Custom `"rpc"` tag for translating struct field names

## license

MIT


CHANGELOG.md
# Changelog

## 1.0.0

* Support server method aliases
* Support content encoding
* Custom `"rpc"` tag support
* Extended `iso8601` date and boolean values `true` and `false`


xml/codec_test.go
package xml

import (
	"bytes"
	"encoding/xml"
	"fmt"
	"reflect"
	"testing"
	"time"
)

var (
	testData = map[string]interface{}{
		// boolean
		"<boolean>1</boolean>": true,
		// numbers
		"<int>-5</int>":          -5,
		"<double>1.201</double>": 1.2010,
		// string
		"<string>hello</string>":                   "hello",
		"<string>&lt;&gt;&amp;&#34;&#39;</string>": `<>&"'`,
		// empty array
		"<array><data></data></array>": []interface{}{},
		// array
		"<array><data><value><int>1</int></value><value><int>2</int></value></data></array>": []int{1, 2},
		// nested array
		"<array><data><value><array><data><value><int>1</int></value><value><int>2</int></value></data></array></value></data></array>": []interface{}{[]int{1, 2}},
		// base64
		"<base64>aGVsbG8=</base64>": []byte("hello"),
		// datetime
		"<dateTime.iso8601>20040101T12:30:10</dateTime.iso8601>": time.Date(2004, time.January, 1, 12, 30, 10, 0, time.UTC),
		// empty struct
		"<struct></struct>": struct{}{},
		// struct
		"<struct><member><name>firstname</name><value><string>Kofi</string></value></member></struct>": map[string]interface{}{
			"firstname": "Kofi",
		},
		"<struct><member><name>age</name><value><int>10</int></value></member></struct>": struct {
			Number int `rpc:"age"`
		}{Number: 10},
	}

	emptyDataFixtures = []string{
		"",
		"<array><data></data></array>",
		"<struct></struct>",
	}
)

type person struct {
	Name string `rpc:"name"`
	Age  int    `rpc:"age"`
}

func Test_ReadWriteFixtures(t *testing.T) {
	for res, v := range testData {
		valType := reflect.TypeOf(v)
		xval := fmt.Sprintf("<value>%s</value>", res)
		b := bytes.NewBufferString("")
		withCodec(func(c *Codec) error {
			encoded := makeValue(v)

			if err := c.writeRPC(b, v); err != nil {
				assertOk(t, false, err, "encoding error. ", valType)
			}
			assertEqual(t, xval, b.String(), "encoding ", valType)

			decoded := makeValue("")
			if err := c.readRPC(b, &decoded); err != nil {
				assertOk(t, false, "readRPC value with type '", valType, "' ", err)
			}
			assertEqual(t, encoded, decoded, "decoding ", valType)
			return nil
		})
	}
}

func Test_EmptyValues(t *testing.T) {
	withCodec(func(c *Codec) error {
		buf := bytes.NewBufferString("")
		for _, res := range emptyDataFixtures {
			xmlstr := fmt.Sprintf("<value>%s</value>", res)
			buf.WriteString(xmlstr)
			var rpc rpcValue
			c.readRPC(buf, &rpc)
			assertOk(t, rpc.isEmpty(), "empty: ", xmlstr)
		}
		return nil
	})
}

// pipeEncodeDecode encode in and decode result to out
func pipeEncodeDecode(t *testing.T, in interface{}, out interface{}) {
	b := bytes.NewBufferString("")
	withCodec(func(c *Codec) error {
		if err := c.writeRPC(b, in); err != nil {
			assertOk(t, false, err)
		}
		if err := c.readRPC(b, out); err != nil {
			assertOk(t, false, err)
		}
		return nil
	})
}

func Test_ReadwriteValues(t *testing.T) {
	// bool
	var b bool
	pipeEncodeDecode(t, true, &b)
	assertEqual(t, true, b, "bool")

	// string
	var s string
	pipeEncodeDecode(t, "hello", &s)
	assertEqual(t, "hello", s, "string")

	// int
	var n int
	pipeEncodeDecode(t, 20, &n)
	assertEqual(t, 20, n, "int")

	// float
	var d float64
	pipeEncodeDecode(t, 3.142, &d)
	assertEqual(t, 3.142, d, "float")

	// time
	testTime := time.Date(2005, 12, 24, 3, 30, 5, 0, time.UTC)
	var dt time.Time
	pipeEncodeDecode(t, testTime, &dt)
	assertEqual(t, testTime, dt, "time")

	// base64
	b64In := []byte("hello")
	var b64Out []byte
	pipeEncodeDecode(t, b64In, &b64Out)
	assertEqual(t, b64In, b64Out, "base64")

	// slice of ints
	intSliceIn := []int{3, 2}
	var intSliceOut []int
	pipeEncodeDecode(t, intSliceIn, &intSliceOut)
	assertEqual(t, intSliceIn, intSliceOut, "slice of ints")

	// slice of strings
	stringSliceIn := []string{"one", "two"}
	var stringSliceOut []string
	pipeEncodeDecode(t, stringSliceIn, &stringSliceOut)
	assertEqual(t, stringSliceIn, stringSliceOut, "slice of strings")

	// struct
	p1 := person{Name: "Gyedu", Age: 35}
	var p2 person
	pipeEncodeDecode(t, p1, &p2)
	assertEqual(t, p1, p2, "struct")

	// slice of structs
	p3 := []person{person{Name: "Roseline", Age: 35}, person{Name: "Odame", Age: 25}}
	var p4 []person
	pipeEncodeDecode(t, p3, &p4)
	assertEqual(t, p3, p4, "slice of structs")

	// struct with slice of struct
	type friend struct {
		Name    string
		Best    person
		Friends []friend
	}

	p5 := friend{
		Name: "Kofi",
		Best: person{Name: "Richard"},
		Friends: []friend{
			friend{
				Name:    "Roseline",
				Friends: []friend{friend{Name: "Odame"}},
			},
		},
	}
	var p6 friend
	pipeEncodeDecode(t, p5, &p6)
	assertEqual(t, p5, p6, "struct with slice of structs")

	// fault
	f1 := InternalError.New("panic")
	var f2 Fault
	pipeEncodeDecode(t, f1, &f2)
	assertEqual(t, f1, f2, "fault message")
}

func Test_ReadWriteRequest(t *testing.T) {
	b := bytes.NewBufferString("")
	body := person{Name: "Nana", Age: 10}
	withCodec(func(c *Codec) error {
		if err := c.writeRequest(b, "service.Do", body); err != nil {
			assertOk(t, false, "encode request. ", err)
		}
		res := xml.Header + "<methodCall><methodName>service.Do</methodName><params><param>" +
			"<value><struct><member><name>name</name><value><string>Nana</string></value></member>" +
			"<member><name>age</name><value><int>10</int></value></member></struct></value>" +
			"</param></params></methodCall>"
		assertEqual(t, res, b.String(), "encode request")

		var method string
		var p person
		if err := c.readRequest(b, &method, &p); err != nil {
			assertOk(t, false, "decode call. ", err)
		}
		assertEqual(t, "service.Do", method, "decode call method")
		assertEqual(t, body, p, "decode call body")
		return nil
	})
}

func Test_ReadWriteResponse(t *testing.T) {
	b := bytes.NewBufferString("")
	encoded := person{Name: "Nana", Age: 10}
	withCodec(func(c *Codec) error {
		if err := c.writeResponse(b, encoded); err != nil {
			assertOk(t, false, "encoding response. ", err)
		}
		res := xml.Header + "<methodResponse><params><param>" +
			"<value><struct><member><name>name</name><value><string>Nana</string></value></member>" +
			"<member><name>age</name><value><int>10</int></value></member></struct></value>" +
			"</param></params></methodResponse>"
		assertEqual(t, res, b.String(), "encode response")

		var decoded person
		c.readResponse(b, &decoded)
		assertEqual(t, encoded, decoded, "decode response")
		return nil
	})
}

func Test_ReadWriteFault(t *testing.T) {
	b := bytes.NewBufferString("")
	encoded := InternalError.New("error decoding value")
	withCodec(func(c *Codec) error {
		if err := c.writeResponse(b, encoded); err != nil {
			assertOk(t, false, "encode fault. ", err)
		}
		res := xml.Header + "<methodResponse><fault>" +
			"<value><struct><member><name>faultCode</name><value><int>-32603</int></value></member>" +
			"<member><name>faultString</name><value><string>error decoding value</string></value></member></struct></value>" +
			"</fault></methodResponse>"
		assertEqual(t, res, b.String(), "encode fault")

		var params bool
		decoded := c.readResponse(b, &params /*error expected*/)
		assertEqual(t, encoded, decoded, "decode fault")
		return nil
	})
}


xml/assert_test.go
package xml

import (
	"fmt"
	"reflect"
	"testing"
)

func assertOk(tb testing.TB, ok bool, v ...interface{}) {
	if ok {
		tb.Logf("pass: %s", fmt.Sprint(v...))
	} else {
		tb.Fatalf("fail: %s", fmt.Sprint(v...))
	}
}

func assertEqual(tb testing.TB, expected, actual interface{}, v ...interface{}) {
	if reflect.DeepEqual(expected, actual) {
		tb.Logf("pass: %s", fmt.Sprint(v...))
	} else {
		tb.Fatalf("fail: %s \nexpected: %#v \nactual: %#v", fmt.Sprint(v...), expected, actual)
	}
}

func assertNotEqual(tb testing.TB, expected, actual interface{}, v ...interface{}) {
	if !reflect.DeepEqual(expected, actual) {
		tb.Logf("pass: %s", fmt.Sprint(v...))
	} else {
		tb.Fatalf("fail: %s", fmt.Sprint(v...))
	}
}


xml/server_test.go
package xml

import (
	"context"
	"fmt"
	"net/http"
	"runtime"
	"testing"

	"github.com/gorilla/rpc/v2"
)

type PositionalArgs []interface{}
type NumericArgs []int

type Args struct {
	A, B int
}

type Reply struct {
	C int
}

type Arith int

func (t *Arith) Add(r *http.Request, args *Args, reply *Reply) error {
	reply.C = args.A + args.B
	return nil
}

func (t *Arith) Mul(r *http.Request, args *Args, reply *Reply) error {
	reply.C = args.A * args.B
	return nil
}

func (t *Arith) Div(r *http.Request, args *Args, reply *Reply) error {
	if args.B == 0 {
		return InvalidParams.New("divide by zero")
	}
	reply.C = args.A / args.B
	return nil
}

func (t *Arith) Max(r *http.Request, args *NumericArgs, reply *Reply) error {
	params := *args
	if len(params) == 0 {
		reply.C = 0
		return nil
	}

	reply.C = params[0]
	for i := 1; i < len(params); i++ {
		if reply.C < params[i] {
			reply.C = params[i]
		}
	}
	return nil
}

func (t *Arith) Count(r *http.Request, args *PositionalArgs, reply *Reply) error {
	params := *args
	reply.C = len(params)
	return nil
}

func createConn() (*http.Server, *Client) {
	address := "127.0.0.1:5000"
	codec := NewServerCodec()
	codec.RegisterAlias("add", "Add")

	s := rpc.NewServer()
	s.RegisterCodec(codec, "text/xml")
	s.RegisterService(new(Arith), "Arith")

	header := make(http.Header)
	header.Set("Accept-Encoding", "deflate")
	client := NewClient(fmt.Sprintf("http://%s", address), WithHTTPHeader(header))
	return &http.Server{Addr: address, Handler: s}, client
}

func Test_ClientServer(t *testing.T) {
	server, c := createConn()
	defer server.Shutdown(context.Background())
	go server.ListenAndServe()
	runtime.Gosched()

	args := Args{A: 3, B: 3}
	var reply Reply

	err := c.Call("Arith.Add", &reply, args)
	assertEqual(t, nil, err, "Add no error")
	assertEqual(t, 6, reply.C, "Add")

	reply.C = -1
	c.Call("Arith.add", &reply, args)
	assertEqual(t, 6, reply.C, "add lowercase alias")

	c.Call("Arith.Mul", &reply, args)
	assertEqual(t, 9, reply.C, "Mul")

	c.Call("Arith.Div", &reply, args)
	assertEqual(t, 1, reply.C, "Div")

	reply.C = -1
	args = Args{A: 1, B: 0}
	err = c.Call("Arith.Div", &reply, args)
	assertNotEqual(t, nil, err, "error on divide by zero")
	assertEqual(t, -1, reply.C, "fail divide by zero")

	fault, ok := err.(Fault)
	assertOk(t, ok, "expect fault")
	assertEqual(t, int(InvalidParams), fault.Code, "read fault code")
	assertEqual(t, "divide by zero", fault.Message, "read fault string")

	reply.C = -1
	c.Call("Arith.Max", &reply, 5, 9, 7, 5)
	assertEqual(t, 9, reply.C, "Max with multiple homogeneous values")

	reply.C = -1
	c.Call("Arith.Count", &reply, 5, "hello", 7.0, []int{1})
	assertEqual(t, 4, reply.C, "Count with multiple heterogeneous values")

	reply.C = -1
	err = c.Call("Arith.Factorize", &reply, args)
	fault, ok = err.(Fault)
	assertNotEqual(t, nil, err, "error for unknown method")
	assertEqual(t, int(MethodNotFound), fault.Code, "method not found")
}


xml/codec.go
package xml

import (
	"encoding/xml"
	"io"
	"io/ioutil"
	"reflect"
	"strings"
	"sync"
)

var (
	// a pool of codecs for the client/server. use via the withCodec function
	codecPool = &sync.Pool{
		New: func() interface{} { return newCodec() },
	}
	emptyReader = strings.NewReader("")
)

// Codec reads and writes XML-RPC messages.
type Codec struct {
	rd *xmlReader
	wr *xmlWriter
}

// withCodec acquires a codec from a pool for the callback and release when done.
// The callback function should not hold a reference to the codec when it completes.
func withCodec(f func(*Codec) error) error {
	c := codecPool.Get().(*Codec)
	err := f(c)
	codecPool.Put(c)
	return err
}

// newCodec return an XML-RPC codec for reading/writing requests and responses
func newCodec() *Codec {
	return &Codec{
		rd: newReader(emptyReader),
		wr: newWriter(ioutil.Discard),
	}
}

// writeRequest serialzes and writes an XML-RPC methodCall
func (c *Codec) writeRequest(w io.Writer, method string, params ...interface{}) error {
	return c.writeRPC(w, makeCall(method, params...))
}

// writeResponse serialzes and writes value as valid XML-RPC methodResponse
func (c *Codec) writeResponse(w io.Writer, params interface{}) error {
	return c.writeRPC(w, makeResponse(params))
}

// writeRPC serialize a value as XML-RPC
func (c *Codec) writeRPC(w io.Writer, rpc interface{}) error {
	c.wr.reset(w)
	var err error
	switch v := rpc.(type) {
	case methodCall:
		err = c.wr.writeCall(v)
	case methodResponse:
		err = c.wr.writeResponse(v)
	case rpcValue:
		err = c.wr.writeValue(v)
	default:
		err = c.wr.writeValue(makeValue(rpc))
	}
	c.wr.Flush()
	return err
}

// readRequest deserialize an XML-RPC methodCall into the method and params pointer receivers
func (c *Codec) readRequest(r io.Reader, method *string, params interface{}) error {
	if err := checkPointer(params); err != nil {
		return err
	}

	var call methodCall
	if err := c.readRPC(r, &call); err != nil {
		return err
	}
	if call.Method == "" {
		return InvalidRequest.New("invalid method name '%s'", call.Method)
	}
	*method = call.Method
	return call.rpcParams.writeTo(params)
}

// readResponse deserialize an XML-RPC methodResponse into the params pointer receiver.
// If the response returned a Fault, the error will be of type xmlrpc.Error
func (c *Codec) readResponse(r io.Reader, reply interface{}) error {
	if err := checkPointer(reply); err != nil {
		return err
	}

	var res methodResponse
	if err := c.readRPC(r, &res); err != nil {
		return err
	}

	if !res.Fault.isEmpty() {
		var fault Fault
		if err := res.Fault.writeTo(&fault); err != nil {
			return err
		}
		return fault
	}

	return res.rpcParams.writeTo(reply)
}

// readRPC deserialize a valid XML-RPC input
func (c *Codec) readRPC(r io.Reader, value interface{}) error {
	if err := checkPointer(value); err != nil {
		return err
	}

	c.rd.reset(r)
	var err error
	switch v := value.(type) {
	case *methodCall:
		err = c.rd.readCall(v)
	case *methodResponse:
		err = c.rd.readResponse(v)
	case *rpcValue:
		err = c.rd.readValue(v)
	default:
		var rpc rpcValue
		if err = c.rd.readValue(&rpc); err == nil || err == io.EOF {
			err = rpc.writeTo(value)
		}
	}

	if err == nil || err == io.EOF {
		return nil
	}

	if v, ok := err.(*xml.SyntaxError); ok {
		return MalformedInput.New(v.Error())
	}

	return err
}

/// Helper methods ///

// checkPointer validates that the value is a pointer type
func checkPointer(v interface{}) error {
	refPtrKind := reflect.TypeOf(v).Kind()
	if refPtrKind != reflect.Ptr {
		return InternalError.New("error decoding to value. expected pointer got '%s'", refPtrKind)
	}
	return nil
}


xml/fault.go
package xml

import (
	"fmt"
	"strconv"
)

// Fault represents an XML-RPC fault.
type Fault struct {
	Code    int    `rpc:"faultCode"`
	Message string `rpc:"faultString"`
}

// Error returns a formatted error string
func (f Fault) Error() string {
	return fmt.Sprintf("%d: %s", f.Code, f.Message)
}

type faultCode int

// Codes: http://xmlrpc-epi.sourceforge.net/specs/rfc.fault_codes.php
const (
	// parse error
	MalformedInput      faultCode = -32700
	UnsupportedEncoding faultCode = -32701
	InvalidCharacter    faultCode = -32702
	// server error
	InvalidRequest faultCode = -32600
	MethodNotFound faultCode = -32601
	InvalidParams  faultCode = -32602
	InternalError  faultCode = -32603
)

var (
	faultMessages = map[faultCode]string{
		MalformedInput:      "malformed input",
		UnsupportedEncoding: "unsupported encoding",
		InvalidCharacter:    "invalid character for encoding",
		InvalidRequest:      "invalid xml-rpc. not conforming to spec",
		MethodNotFound:      "requested method not found",
		InvalidParams:       "invalid method parameters",
		InternalError:       "internal xml-rpc error",
	}
)

func (f faultCode) String() string {
	return faultMessages[f]
}

func (f faultCode) Error() string {
	return strconv.Itoa(int(f)) + ": " + f.String()
}

func (f faultCode) New(format string, v ...interface{}) Fault {
	s := fmt.Sprintf(format, v...)
	if len(s) == 0 {
		s = f.String()
	}
	return Fault{Code: int(f), Message: s}
}


xml/client.go
package xml

import (
	"bytes"
	"net/http"
	"sync"
)

// A Client is used to make XML-RPC calls.
type Client struct {
	url        string
	username   string
	password   string
	client     *http.Client
	header     http.Header
	bufPoolMap map[string]*sync.Pool
	bufMtx     sync.Mutex
}

// NewClient returns a new XML-RPC client.
func NewClient(url string, options ...func(*Client)) *Client {
	c := &Client{
		url:        url,
		bufPoolMap: make(map[string]*sync.Pool),
		client:     http.DefaultClient,
		header:     make(http.Header),
	}

	for _, opt := range options {
		opt(c)
	}

	c.header.Set("Content-Type", "text/xml")

	return c
}

// WithBasicAuth configure client with basic HTTP authentication.
func WithBasicAuth(username, password string) func(*Client) {
	return func(c *Client) {
		c.username = username
		c.password = password
	}
}

// WithHTTPClient confgure a custom HTTP client to use for connecting to server.
func WithHTTPClient(httpClient *http.Client) func(*Client) {
	return func(c *Client) {
		c.client = httpClient
	}
}

// WithHTTPHeader configure headers to add to each request.
func WithHTTPHeader(header http.Header) func(*Client) {
	return func(c *Client) {
		for k, v := range header {
			for _, w := range v {
				c.header.Set(k, w)
			}
		}
	}
}

// Call sends an XML-RPC request to the server.
// If a non-nil error is returned, it may be an rpc.Fault or some other type of error
func (c *Client) Call(method string, reply interface{}, args ...interface{}) error {
	return withCodec(func(codec *Codec) error {
		return c.withBuffer(method, func(buf *bytes.Buffer) error {
			if err := codec.writeRequest(buf, method, args...); err != nil {
				return err
			}

			req, err := http.NewRequest("POST", c.url, buf)
			if err != nil {
				return err
			}

			// set custom request headers
			req.Header = c.header

			if c.username != "" && c.password != "" {
				req.SetBasicAuth(c.username, c.password)
			}

			resp, err := c.client.Do(req)
			if err != nil {
				return err
			}

			dec := newDecompressor(resp)
			err = codec.readResponse(dec, reply)
			dec.Close()
			return err
		})
	})
}

func (c *Client) withBuffer(method string, fn func(*bytes.Buffer) error) error {
	c.bufMtx.Lock()
	pool, ok := c.bufPoolMap[method]
	if !ok {
		pool = &sync.Pool{
			New: func() interface{} { return bytes.NewBuffer([]byte{}) },
		}
		c.bufPoolMap[method] = pool
	}
	c.bufMtx.Unlock()

	buf := pool.Get().(*bytes.Buffer)
	err := fn(buf)
	buf.Reset()
	pool.Put(buf)
	return err
}


xml/reader.go
package xml

import (
	"encoding/base64"
	"encoding/xml"
	"fmt"
	"io"
	"strconv"
	"strings"
	"time"
)

const (
	iso8601         = "20060102T15:04:05"
	rfc3339NoTZ     = "2006-01-02T15:04:05"
	rfc3339HyphenTZ = "2006-01-02T15:04:05-07:00"
)

var (
	dateTimeFormats = [4]string{iso8601, time.RFC3339, rfc3339HyphenTZ, rfc3339NoTZ}
	boolDecodeMap   = map[string]bool{"1": true, "true": true, "0": false, "false": false}
	valueTagSet     = map[string]bool{}
)

// reads an XML-RPC input from an io.Reader
type xmlReader struct {
	dec  *xml.Decoder // for XML pull parsing
	peek xml.Token    // next token we peeked
}

func init() {
	for _, t := range [8]xmlTag{stringTag, intTag, base64Tag, dateTimeTag, doubleTag, booleanTag, arrayTag, structTag} {
		valueTagSet[tagNames[t]] = true
	}
	valueTagSet["i4"] = true //alternative for int tags
}

func newReader(r io.Reader) *xmlReader {
	return &xmlReader{
		dec: xml.NewDecoder(r),
	}
}

// resets the reader internal state
func (r *xmlReader) reset(rd io.Reader) {
	r.peek = nil
	r.dec = xml.NewDecoder(rd)
}

func (r *xmlReader) readHeader() error {
	r.trim()
	t, err := r.token()
	if err != nil {
		return err
	}
	if hdr, ok := t.(xml.ProcInst); ok {
		if hdr.Target == "xml" {
			return nil
		}
	}
	r.putToken(t)
	return nil
}

func (r *xmlReader) readCall(rpc *methodCall) error {
	if err := r.readHeader(); err != nil {
		return err
	}

	err := r.expectStart("methodCall")
	if err != nil {
		return err
	}

	if err = r.expectStart("methodName"); err != nil {
		return err
	}
	if rpc.Method, err = r.nextText(); err != nil {
		return err
	}
	if err = r.expectEnd("methodName"); err != nil {
		return err
	}

	if err = r.readParams(&rpc.rpcParams); err != nil {
		return err
	}

	return r.expectEnd("methodCall")
}

func (r *xmlReader) readResponse(rpc *methodResponse) error {
	if err := r.readHeader(); err != nil {
		return err
	}

	err := r.expectStart("methodResponse")
	if err != nil {
		return err
	}

	if err = r.readParams(&rpc.rpcParams); err != nil && err != io.EOF {
		if err = r.expectStart("fault"); err != nil {
			return err
		}
		if err = r.readValue(&rpc.Fault); err != nil {
			return err
		}
		if err = r.expectEnd("fault"); err != nil {
			return err
		}
	}
	return r.expectEnd("methodResponse")
}

func (r *xmlReader) readParams(rpc *rpcParams) error {
	err := r.expectStart("params")
	if err != nil {
		return err
	}

	for {
		err = r.expectStart("param")
		if err != nil {
			// we assume we hit the end
			break
		}

		var v rpcValue
		if err = r.readValue(&v); err != nil {
			return err
		}
		rpc.Params = append(rpc.Params, v)

		err = r.expectEnd("param")
		if err != nil {
			return err
		}
	}

	return r.expectEnd("params")
}

// readValue decodes and reads the next value
func (r *xmlReader) readValue(rpc *rpcValue) error {
	err := r.expectStart("value")
	if err != nil {
		return err
	}

	// determine the type of value
	se, err := r.nextStart()
	if err != nil {

		// empty value or unwrapped string
		s, err := r.nextText()
		if err != nil {
			return r.expectEnd("value")
		}
		// treat empty value as empty string
		rpc.value = s
		rpc.kind = stringKind
		return err
	}

	if !valueTagSet[se.Name.Local] {
		return fmt.Errorf("parsing error. expected valid rpc value element got '%s'", se.Name.Local)
	}

	r.putToken(se)

	switch se.Name.Local {
	case "array":
		err = r.readArray(rpc)
	case "struct":
		err = r.readStruct(rpc)
	default:
		err = r.readPrimitive(rpc)
	}

	if err != nil {
		return err
	}

	// match end tag
	return r.expectEnd("value")
}

// readPrimitive reads the next primitive value
func (r *xmlReader) readPrimitive(rpc *rpcValue) error {
	// assume start is valid since we always come via readValue()
	se, err := r.nextStart()
	if err != nil {
		return err
	}

	s, _ := r.nextText()
	if err = r.expectEnd(se.Name.Local); err != nil {
		return err
	}

	var ok bool

	switch se.Name.Local {
	case "string":
		rpc.value = s
		rpc.kind = stringKind
	case "boolean":
		if rpc.value, ok = boolDecodeMap[s]; !ok {
			return InvalidRequest.New("error writing boolean '%s'", s)
		}
		rpc.kind = booleanKind
	case "int", "i4":
		if rpc.value, err = strconv.Atoi(s); err != nil {
			return InvalidRequest.New("error writing int '%s'", s)
		}
		rpc.kind = intKind
	case "double":
		if rpc.value, err = strconv.ParseFloat(s, 64); err != nil {
			return InvalidRequest.New("error writing double '%s'", s)
		}
		rpc.kind = doubleKind
	case "base64":
		rpc.value, err = base64.StdEncoding.DecodeString(s)
		rpc.kind = base64Kind
	case "dateTime.iso8601":
		for _, dateFmt := range dateTimeFormats {
			if rpc.value, err = time.Parse(dateFmt, s); err == nil {
				break
			}
		}
		rpc.kind = dateTimeKind
	default:
		return fmt.Errorf("unhandled tag. '%s'", se.Name.Local)
	}
	return err
}

// readArray reads an array value
func (r *xmlReader) readArray(rpc *rpcValue) error {
	r.nextStart() // <array>

	// <data>
	err := r.expectStart("data")
	if err != nil {
		return err
	}

	var array []rpcValue

	for {
		se, err := r.nextStart()
		if err != nil {
			// empty array is allowed although just a waste of bytes
			break
		}

		// we expect every start element to be a value
		if se.Name.Local != "value" {
			return fmt.Errorf("parsing error. invalid element '%s'", se.Name.Local)
		}

		// read the values
		var val rpcValue

		r.putToken(se)
		if err := r.readValue(&val); err != nil {
			return err
		}

		array = append(array, val)
	}

	rpc.value = array
	rpc.kind = arrayKind

	err = r.expectEnd("data")
	if err != nil {
		return err
	}

	return r.expectEnd("array")
}

// readStruct reads the struct value
func (r *xmlReader) readStruct(rpc *rpcValue) error {
	r.nextStart() // <struct>

	var members []rpcEntry

	for {
		err := r.expectStart("member")
		if err != nil {
			// empty structs are allowed
			break
		}

		// read the member details
		var entry rpcEntry

		// start member
		if err = r.expectStart("name"); err != nil {
			return err
		}
		if entry.Name, err = r.nextText(); err != nil {
			return err
		}
		if err = r.expectEnd("name"); err != nil {
			return err
		}

		if err = r.readValue(&entry.Value); err != nil {
			return err
		}

		if err = r.expectEnd("member"); err != nil {
			return err
		}

		members = append(members, entry)
	}

	rpc.value = members
	rpc.kind = structKind

	// we expect to be at the end of the struct
	return r.expectEnd("struct")
}

// nextText read the required next token as text. treat empty text as an error
func (r *xmlReader) nextText() (string, error) {
	t, err := r.token()
	if t == nil {
		return "", err
	}
	if cd, ok := t.(xml.CharData); ok {
		return string(cd), nil
	}
	r.putToken(t)
	return "", fmt.Errorf("expected chardata but got '%#v'", t)
}

// nextStart return the next token expected as an xml.StartElement
func (r *xmlReader) nextStart() (xml.StartElement, error) {
	r.trim()
	t, err := r.token()
	if t == nil {
		return xml.StartElement{}, err
	}
	if se, ok := t.(xml.StartElement); ok {
		return se, nil
	}
	r.putToken(t)
	return xml.StartElement{}, fmt.Errorf("expected start element but got '%s'", t)
}

// nextEnd return the next token expected as an xml.EndElement
func (r *xmlReader) nextEnd() (xml.EndElement, error) {
	r.trim()
	t, err := r.token()
	if t == nil {
		return xml.EndElement{}, err
	}
	if end, ok := t.(xml.EndElement); ok {
		return end, nil
	}
	r.putToken(t)
	return xml.EndElement{}, fmt.Errorf("expected end element but got '%s'", t)
}

// expect a start element with the given name
func (r *xmlReader) expectStart(name string) error {
	se, err := r.nextStart()
	if err != nil {
		return err
	}
	if se.Name.Local != name {
		r.putToken(se)
		return fmt.Errorf("parsing error. expected start element '%s' but got '%s'", name, se.Name.Local)
	}
	return nil
}

// expect an end element with the given name
func (r *xmlReader) expectEnd(name string) error {
	end, err := r.nextEnd()
	if err != nil {
		return err
	}
	if end.Name.Local != name {
		r.putToken(end)
		return fmt.Errorf("parsing error. expected end element '%s' but got '%s'", name, end.Name.Local)
	}
	return nil
}

// token returns the next token from the XML stream
func (r *xmlReader) token() (xml.Token, error) {
	if r.peek != nil {
		t := r.peek
		r.peek = nil
		return t, nil
	}
	return r.dec.RawToken()
}

func (r *xmlReader) trim() {
	for {
		t, _ := r.token()
		if cd, ok := t.(xml.CharData); ok {
			if strings.TrimSpace(string(cd)) == "" {
				continue
			}
		}
		r.putToken(t)
		break
	}
}

// pushes the token on the top of decoding stream
func (r *xmlReader) putToken(t xml.Token) {
	r.peek = t
}
